File: domain-driven-design/csharp/src/DDD.Domain/Orders/OrderStatus.cs
namespace DDD.Domain.Orders;

using DDD.Domain.Common;

public class OrderStatus : Enumeration
{
    public static OrderStatus Pending = new(1, nameof(Pending));
    public static OrderStatus Processing = new(2, nameof(Processing));
    public static OrderStatus Shipped = new(3, nameof(Shipped));
    public static OrderStatus Delivered = new(4, nameof(Delivered));
    public static OrderStatus Cancelled = new(5, nameof(Cancelled));
    
    public OrderStatus(int id, string name) : base(id, name)
    {
    }
}

File: domain-driven-design/java/src/com/example/domain/model/valueobject/OrderStatus.java
package com.example.domain.model.valueobject;

/**
 * 注文の状態を表す列挙型
 * DDDでは列挙型も値オブジェクトの一種として扱うことができる
 */
public enum OrderStatus {
    CREATED("作成済み"),
    PENDING_PAYMENT("支払い待ち"),
    PAID("支払い完了"),
    PROCESSING("処理中"),
    SHIPPED("発送済み"),
    DELIVERED("配達完了"),
    CANCELLED("キャンセル済み"),
    RETURNED("返品済み");

    private final String displayName;

    OrderStatus(String displayName) {
        this.displayName = displayName;
    }

    public String getDisplayName() {
        return displayName;
    }

    public boolean isEditable() {
        return this == CREATED || this == PENDING_PAYMENT;
    }

    public boolean isCancellable() {
        return this == CREATED || this == PENDING_PAYMENT || this == PAID || this == PROCESSING;
    }

    public boolean isShippable() {
        return this == PAID || this == PROCESSING;
    }

    public boolean isFinalized() {
        return this == DELIVERED || this == CANCELLED || this == RETURNED;
    }
}